### ISSUE TITLE
Scalelite 1.3.1 - nginx: [emerg] unknown "scalelite_api_port" variable

### ISSUE THREAD

**Issue by Bevito** [blindsidenetworks/scalelite#743]
**Describe the bug**
Since the last update (1.3.1) nginx docker image does not work anymore.
In the system logs (/var/log/syslog) I see : 

nginx: [emerg] unknown "scalelite_api_port" variable

If I add 
SCALELITE_API_PORT=3000
to /etc/default/scalelite I had the error : 
/etc/nginx/start: line 29: =3000: not found

**To Reproduce**
Just update from scalelite 1.3 to 1.3.1 with 
sudo systemctl restart scalelite.target

***Deployment:***
1. Deployment: systemd
2. Versions: 
Docker : Docker version 20.10.12, build e91ed57
Host OS : Debian 11.2

**Additional context**
I tried to roll back with the variable SCALELITE_TAG : 

Changing "SCALELITE_TAG=v1" to "SCALELITE_TAG=v1.3" in /etc/default/scalelite but the problem seems to be the same.
Docker is still using version 1.3.1

Am I doing something wrong ? 

Thanks a lot for any help

**Comment by SiJux** [comment]
Same issue here.
I've done some debugging and I think I've found the Problem:
https://github.com/blindsidenetworks/scalelite/blob/e9dbe8c9ce87ec6a254322535266873bd97a43ee/nginx/start#L29

I think it should be:
SCALELITE_API_PORT=${SCALELITE_API_PORT:-3000}
not
$SCALELITE_API_PORT=${SCALELITE_API_PORT:-3000}

A short-term-fix would be wonderful. 

Regards Simon

**Comment by SiJux** [comment]
I think there is also a problem with the variable-replacement for SCALELITE_API_PORT.

**Comment by SiJux** [comment]
Here is a workaround:
Just on your scalelite-host

```
cd /tmp/
wget https://raw.githubusercontent.com/blindsidenetworks/scalelite/master/nginx/start
chmod 777 start
```

Modify start as follow:

```[root@scalelite tmp]# diff -Naur start start.orig
--- start       2021-12-24 05:43:41.749728245 +0100
+++ start.orig  2021-12-24 08:24:51.985783699 +0100
@@ -26,9 +26,8 @@
 fi

 #apply SCALELITE_API port from environment variable or use default port 3000
-SCALELITE_API_PORT=${SCALELITE_API_PORT:-3000}
-#envsubst '$SCALELITE_API_PORT' <$nginx_template >/etc/nginx/conf.d/scalelite.conf
-sed -i 's/\$SCALELITE_API_PORT/3000/g' /etc/nginx/conf.d/*
+$SCALELITE_API_PORT=${SCALELITE_API_PORT:-3000}
+envsubst '$SCALELITE_API_PORT' <$nginx_template >/etc/nginx/conf.d/scalelite.conf

 envsubst '$URL_HOST' <$nginx_template >/etc/nginx/conf.d/scalelite.conf
 unset nginx_template
```

Add the following `SCALELITE_NGINX_EXTRA_OPTS ` to `/etc/default/scalelite`:

```--mount type=bind,source=/tmp/start,target=/etc/nginx/start```

Now restart `scalelite-nginx`

**Comment by Bevito** [comment]
Thank you very much Simon, your workaround help me a lot.
In my case, I had to modify SCALELITE_NGINX_EXTRA_OPTS like this : 
SCALELITE_NGINX_EXTRA_OPTS=--mount type=bind,source=/etc/sectigo,target=/etc/nginx/ssl,readonly \--mount type=bind,source=/tmp/start,target=/etc/nginx/start

Thanks again !

**Comment by SiJux** [comment]
Great!

Merry Christmas to you!

**Comment by Uffi64** [comment]
I tried to apply the patch, but it didn't work out for me completely.
So I  fiddeld around a bit until it worked for me in the end. It's surely not the most elegant solution, but what the heck... it's christmas day and I have better things to do today ;-)

Here's what's in my `start` script now, in case someon wants to know:

```
  :
fi

# apply SCALELITE_API port from environment variable or use default port 3000
SCALELITE_API_PORT=${SCALELITE_API_PORT:-3000}
#envsubst '$SCALELITE_API_PORT' <$nginx_template >/etc/nginx/conf.d/scalelite.conf
cp $nginx_template /etc/nginx/conf.d/scalelite.conf.workaround
sed -i 's/\$SCALELITE_API_PORT/3000/g' /etc/nginx/conf.d/scalelite.conf.workaround
cp /etc/nginx/conf.d/scalelite-ssl.template /etc/nginx/conf.d/scalelite.conf

envsubst '$URL_HOST' </etc/nginx/conf.d/scalelite.conf.workaround >/etc/nginx/conf.d/scalelite.conf
unset nginx_template

mkdir -p /run/nginx
  :
```

In my `/etc/default/scalelite` it's now:

```
SCALELITE_NGINX_EXTRA_OPTS=--mount type=bind,source=/etc/letsencrypt,target=/etc/nginx/ssl,readonly --mount type=bind,source=/etc/bigbluebutton/nginx,target=/etc/nginx/conf.d/scalelite --mount type=bind,source=/tmp/start,target=/etc/nginx/start
```

@SiJux  
Thx a lot for your post. It saved the rest of the day for me. 

Merry christmas everyone!!!

Uffi

**Comment by krim404** [comment]
This is especially problematic, as it breaks existing 1.3 systems over christmas because the 1.3 image also got overwritten with this faulty revision.

So restarting the server will break all 1.3 scalelite servers, which just happened today and resulted in an alarm state at christmas. 

there is only a posibility to downgrade to 1.2.0.1 without forking or manipulating the script.

This is a really nasty bug which should be fixed immediately @jfederico

**Comment by flyinghuman** [comment]
> I tried to apply the patch, but it didn't work out for me completely. So I fiddeld around a bit until it worked for me in the end. It's surely not the most elegant solution, but what the heck... it's christmas day and I have better things to do today ;-)
> 
> Here's what's in my `start` script now, in case someon wants to know:
> 
> ```
>   :
> fi
> 
> # apply SCALELITE_API port from environment variable or use default port 3000
> SCALELITE_API_PORT=${SCALELITE_API_PORT:-3000}
> #envsubst '$SCALELITE_API_PORT' <$nginx_template >/etc/nginx/conf.d/scalelite.conf
> cp $nginx_template /etc/nginx/conf.d/scalelite.conf.workaround
> sed -i 's/\$SCALELITE_API_PORT/3000/g' /etc/nginx/conf.d/scalelite.conf.workaround
> 
> envsubst '$URL_HOST' </etc/nginx/conf.d/scalelite.conf.workaround >/etc/nginx/conf.d/scalelite.conf
> unset nginx_template
> 
> mkdir -p /run/nginx
>   :
> ```
> 
> In my `/etc/default/scalelite` it's now:
> 
> ```
> SCALELITE_NGINX_EXTRA_OPTS=--mount type=bind,source=/etc/letsencrypt,target=/etc/nginx/ssl,readonly --mount type=bind,source=/etc/bigbluebutton/nginx,target=/etc/nginx/conf.d/scalelite --mount type=bind,source=/tmp/start,target=/etc/nginx/start
> ```
> 
> @SiJux Thx a lot for your post. It saved the rest of the day for me.
> 
> Merry christmas everyone!!!
> 
> Uffi

this worked for me too... Thanks!

**Comment by Ayildiz89** [comment]
Hi Everyone ,
I also changed versions of Nginxs version , and i'm also getting this error on docker;
nginx: [emerg] unknown "scalelite_api_port" variable
scalelite-nginx                 | Startup script was run as init, re-execing using tini.
scalelite-nginx                 | Generating templated nginx configuration...
scalelite-nginx                 | Using SSL configuration template.
scalelite-nginx                 | /etc/nginx/start: line 29: 3000=3000: not found
scalelite-nginx                 | Starting nginx periodic reload process...

how can i with deal with this problem ?

**Comment by Uffi64** [comment]
@Ayildiz89 hmmm... did you read the posts above and tried these things?

**Comment by Ayildiz89** [comment]
@Uffi64 Yes but maybe i couldn’t set script properly can you send me content of start script? or have you any advise ?

**Comment by dannyplace** [comment]
> @Uffi64 Yes but maybe i couldn’t set script properly can you send me content of start script? or have you any advise ?

Hi there,

Maybe you can post exactly what you are doing? So that we can pinpoint where it goes wrong.

**Comment by Ayildiz89** [comment]
> > @Uffi64 Yes but maybe i couldn’t set script properly can you send me content of start script? or have you any advise ?
> 
> Hi there,
> 
> Maybe you can post exactly what you are doing? So that we can pinpoint where it goes wrong.

Hi,

this is my start script ;

#!/bin/sh

if [ $$ == 1 ]; then
	echo "Startup script was run as init, re-execing using tini."
	exec /sbin/tini -- $0 "$@"
fi

# For backward compatibility, set NGINX_BEHIND_PROXY if not defined but BEHIND_PROXY is.
if [ -z "$NGINX_BEHIND_PROXY" ] && [ -n "$BEHIND_PROXY" ] ; then
  NGINX_BEHIND_PROXY="true"
fi

echo "Generating templated nginx configuration..."
if [ "$NGINX_RECORDINGS_ONLY" = "true" ] ; then
	echo "Using nginx recordings only configuration template."
	nginx_template=/etc/nginx/conf.d/scalelite-recordings-only.template
elif [ "$NGINX_BEHIND_PROXY" = "true" ] ; then
	echo "Using proxy configuration template."
	nginx_template=/etc/nginx/conf.d/scalelite-proxy.template
elif [ "$NGINX_SSL" = "true" ] ; then
	echo "Using SSL configuration template."
	nginx_template=/etc/nginx/conf.d/scalelite-ssl.template
else
	echo "Using non-SSL configuration template."
	nginx_template=/etc/nginx/conf.d/scalelite.template
fi

# apply SCALELITE_API port from environment variable or use default port 3000
SCALELITE_API_PORT=${SCALELITE_API_PORT:-3000}
envsubst '$SCALELITE_API_PORT' <$nginx_template >/etc/nginx/conf.d/scalelite.conf

envsubst '$URL_HOST' <$nginx_template >/etc/nginx/conf.d/scalelite.conf
unset nginx_template

mkdir -p /run/nginx

echo "Starting nginx periodic reload process..."
while :; do
	sleep 6h
	echo "Reloading nginx..."
	nginx -s reload
done &

echo "Starting nginx..."
exec nginx -g 'daemon off;'

**Comment by dannyplace** [comment]
@Ayildiz89 

**Create the start script and place it in /tmp/ or any folder you like:** (you can copy this script)

```
#!/bin/sh

if [ $$ == 1 ]; then
	echo "Startup script was run as init, re-execing using tini."
	exec /sbin/tini -- $0 "$@"
fi

# For backward compatibility, set NGINX_BEHIND_PROXY if not defined but BEHIND_PROXY is.
if [ -z "$NGINX_BEHIND_PROXY" ] && [ -n "$BEHIND_PROXY" ] ; then
  NGINX_BEHIND_PROXY="true"
fi

echo "Generating templated nginx configuration..."
if [ "$NGINX_RECORDINGS_ONLY" = "true" ] ; then
	echo "Using nginx recordings only configuration template."
	nginx_template=/etc/nginx/conf.d/scalelite-recordings-only.template
elif [ "$NGINX_BEHIND_PROXY" = "true" ] ; then
	echo "Using proxy configuration template."
	nginx_template=/etc/nginx/conf.d/scalelite-proxy.template
elif [ "$NGINX_SSL" = "true" ] ; then
	echo "Using SSL configuration template."
	nginx_template=/etc/nginx/conf.d/scalelite-ssl.template
else
	echo "Using non-SSL configuration template."
	nginx_template=/etc/nginx/conf.d/scalelite.template

fi

# apply SCALELITE_API port from environment variable or use default port 3000
SCALELITE_API_PORT=${SCALELITE_API_PORT:-3000}
#envsubst '$SCALELITE_API_PORT' <$nginx_template >/etc/nginx/conf.d/scalelite.conf
cp $nginx_template /etc/nginx/conf.d/scalelite.conf.workaround
sed -i 's/\$SCALELITE_API_PORT/3000/g' /etc/nginx/conf.d/scalelite.conf.workaround

envsubst '$URL_HOST' </etc/nginx/conf.d/scalelite.conf.workaround >/etc/nginx/conf.d/scalelite.conf
unset nginx_template

mkdir -p /run/nginx

echo "Starting nginx periodic reload process..."
while :; do
	sleep 6h
	echo "Reloading nginx..."
	nginx -s reload
done &

echo "Starting nginx..."
exec nginx -g 'daemon off;'
```

Next do a chmod on the file so that the containers can read it.

chmod 777 /tmp/start


**Next edit the: /etc/default/scalelite**

Add or edit:

SCALELITE_NGINX_EXTRA_OPTS=--mount type=bind,source=/tmp/start,target=/etc/nginx/start 

**Restart all the scalelite services:**

sudo systemctl restart scalelite.target

**Comment by dannyplace** [comment]
Just for the reminder, placing the start script in /tmp/ is not the best way to do it because on system reboot usually linux cleares the /tmp/ folder. So in order to make it more permanent place the start file in another folder.

**Comment by Ayildiz89** [comment]
@dannyplace thank you so much , just one thing. All service are running on docker with docker-compose  , where can i find /etc/default/scalelite this folder to change it..? Normally i have .env file so that all container get env. variab. from there. I added this SCALELITE_NGINX_EXTRA_OPTS=--mount type=bind,source=/tmp/start,target=/etc/nginx/start in .env file but i m getting same error...

**Comment by dannyplace** [comment]
> @dannyplace thank you so much , just one thing. All service are running on docker with docker-compose , where can i find /etc/default/scalelite this folder to change it..? Normally i have .env file so that all container get env. variab. from there. I added this SCALELITE_NGINX_EXTRA_OPTS=--mount type=bind,source=/tmp/start,target=/etc/nginx/start in .env file but i m getting same error...

In docker-compose you should have this to read the scalelite environment file (notice the env_file section):

```
scalelite-nginx:
    image: blindsidenetwks/scalelite:master-nginx
    container_name: scalelite-nginx

    env_file:
      - /etc/default/scalelite
```

**Comment by RonZ-dev** [comment]
I think that he wants to mount the start file using docker-compose.
This is possible by using the volumes: part, for example:
```
  scalelite-nginx:
    image: blindsidenetwks/scalelite:master-nginx
    container_name: scalelite-nginx

    env_file:
      - ./data/scalelite
    restart: always

    volumes:
      **- /tmp/start:/etc/nginx/start**
      - ./data/var/bigbluebutton/published:/var/bigbluebutton/published
```
Keep in mind that like @dannyplace stated its not handy to place the start file in /tmp because it will get deleted on reboot.

**Comment by Ayildiz89** [comment]
@dannyplace thank you so much it finally worked :) 👍

**Comment by ywarnier** [comment]
Just reporting here, as a quickfix for a Docker and Scalelite noob, when using Scalelite with SSL from letsencrypt, that the following worked for me:
 - using the start script from https://github.com/blindsidenetworks/scalelite/commit/51b820a837788fccd440aa10cbf0202f425a40ae#diff-2452b8cc6ca931b36195be4db6f5dd9a0da7d12e438249d1c0f3e05f23828d30
 - putting it in a common directory (I saved it as /usr/local/bin/scalelite-alt-start)
 - chmoding it to be executable: `chmod 777 /usr/local/bin/scalelite-alt-start`
 - modifying /etc/default/scalelite to use this script like so: `SCALELITE_NGINX_EXTRA_OPTS=--mount type=bind,source=/etc/letsencrypt,target=/etc/nginx/ssl,readonly --mount type=bind,source=/usr/local/bin/scalelite-alt-start,target=/etc/nginx/start`
 - then restarting the service (I restarted both, not sure if that was necessary): 
 ```sudo systemctl restart scalelite.target
 sudo systemctl restart scalelite-nginx.service
 ```
 Thanks to all in this thread for sharing the issue and efforts to fix it, and happy 2022!

**Comment by farhatahmad** [comment]
This should be fixed now in 1.3.2. If you pull the latest image, everything should be working

**Comment by dannyplace** [comment]
> This should be fixed now in 1.3.2. If you pull the latest image, everything should be working

Yes, I can confirm!
